jumanji/environments/routing/op/env.py

```diff
@@ -0,0 +1,313 @@
+            viewer: 'Viewer' used for rendering. Defaults to the 'OPViewer' with 'human' render mode.   
+        """
+        
+        
+        self.generator = generator or generator.UniformGenerator(
+            num_nodes=20,
+            max_length=2,
+        )
+        self.num_nodes = self.generator.num_nodes
+        self.max_length = self.generator.max_length
+        self.reward_fn = reward_fn or reward.DenseReward()
+        self._viewer = viewer #or OPViewer(name="OP", render_mode="human")
+        
+        
+    def __repr__(self) -> str:
+        return f"OP environment with {self.num_nodes} nodes and travel budget of {self.max_length}."
+    
+    
+    def reset(self, key: chex.PRNGKey) -> Tuple[types_.State, TimeStep[types_.Observation]]:
+           """Reset the environment.
+           
+           Args:
+                Key: used to randomly generate the coordinates.
+                
+                
+           Returns:
+                state: State object corresponding to the new state of the environment.
+                timestep: Timestep object corresponding to the first timestep returned 
+                by the environment.     
+           """
+           state = self.generator(key)
+           timestep = restart(observation=self.__state_to_observation(state))
+           return state, timestep
+       
+       
+    def step(
+        self, state: types_.State, action: chex.Numeric
+    ) -> Tuple[types_.State, TimeStep[types_.Observation]]:
+        """Run one timestep of the environment's dynamics.
+        
+        
+        Args:
+            state: 'State' object containing the dynamics of the environment.
+            action: 'Array' containing the index of the next node to visit.
+            
+            
+        Returns:
+            state: the next state of the environment.
+            timestep: the timestep to be observed.    
+        """
+        # Valid if node has not been visited and/or there is sufficient travel budget
+        is_valid = ~state.visited_mask[action] & (state.length[action] < state.remaining_max_length)
+
+        next_state = jax.lax.cond(
+            is_valid,
+            self._update_state,
+            lambda *_: state,
+            state,
+            action,
+        )
+        
+        reward = self.reward_fn(state, action, next_state, is_valid)
+        observation = self._state_to_observation(next_state)
+        
+        # Terminate if all the travel budget is used up or the action is invalid
+        is_done = (state.remaining_max_length == next_state.length[action]) | ~is_valid
+        
+        timestep = jax.lax.cond(
+            is_done,
+            termination,
+            transition,
+            reward,
+            observation,
+        )
+        return next_state, timestep
+    
+    
+    def observation_spec(self) -> specs.Spec[types_.Observation]:
```

Comment:
Have you also made sure that the observation is exactly the same as the one used in Attention, Learn to solve routing problems?